compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java

```diff
@@ -384,17 +638,73 @@ public void performCodeGen(ModuleContext moduleContext, CompilationCache compila
+        // TODO do the merging inside usedBIRNodeAnalyzer
+        bLangPackage.testablePkgs.get(0).symbol.invocationData = bLangPackage.symbol.invocationData;
+
         String testJarFileName = jarFileName + TEST_JAR_FILE_NAME_SUFFIX;
         CompiledJarFile compiledTestJarFile = jvmCodeGenerator.generateTestModule(bLangPackage.testablePkgs.get(0),
                 isRemoteMgtEnabled);
         try {
             ByteArrayOutputStream byteStream = JarWriter.write(compiledTestJarFile, getAllResources(moduleContext));
-            compilationCache.cachePlatformSpecificLibrary(this, testJarFileName, byteStream);
+            moduleContext.getCompilationCache().cachePlatformSpecificLibrary(this, testJarFileName, byteStream, true);
         } catch (IOException e) {
             throw new ProjectException("Failed to cache generated test jar, module: " + moduleContext.moduleName());
         }
     }
 
+    public void optimizeBirPackage(BPackageSymbol bPackageSymbol) {
+        UsedBIRNodeAnalyzer.InvocationData invocationData = bPackageSymbol.invocationData;
+        BIRNode.BIRPackage birPackage = bPackageSymbol.bir;
+
+        if (!invocationData.moduleIsUsed) {
+            // Thrown if the compiler tries to pack an UNUSED thin JAR to the final executable with --optimize flag
+            throw new IllegalStateException(
+                    String.format(
+                            "BIR Package %s should not be packed to final executable because it is not used!",
+                            bPackageSymbol.getName()));
+        }
+
+        bPackageSymbol.imports.removeIf(pkgSymbol -> pkgSymbol != null && unusedPackageIDs.contains(pkgSymbol.pkgID));
+        birPackage.importModules.removeIf(module -> isUnusedPkgID(module.packageID));
+        birPackage.functions.removeIf(currentFunc -> currentFunc.getUsedState() == UsedState.UNUSED);
+        birPackage.typeDefs.removeIf(typeDef -> typeDef.getUsedState() == UsedState.UNUSED);
+        optimizeImmutableTypeDefs(invocationData, bPackageSymbol.pkgID);
+
+        // FP optimization for functions with default params
+        birPackage.globalVars.removeIf(gVar -> gVar.getUsedState() == UsedState.UNUSED);
+        invocationData.getFpDataPool().forEach(UsedBIRNodeAnalyzer.FunctionPointerData::deleteIfUnused);
+
+        // TODO Attached function optimization with polymorphism handling
+    }
+
+    private boolean isUnusedPkgID(PackageID packageID) {
+        return unusedPackageIDs.stream().anyMatch(unusedPkgID -> unusedPkgID.nameComps.equals(packageID.nameComps));
+    }
+
+    private void optimizeImmutableTypeDefs(UsedBIRNodeAnalyzer.InvocationData invocationData, PackageID pkgID) {
+        HashSet<BIRNode.BIRTypeDefinition> deadTypeDefs = new HashSet<>(invocationData.unusedTypeDefs);
+        Map<SelectivelyImmutableReferenceType, BIntersectionType> immutableTypeMap =
+                symbolTable.immutableTypeMaps.get(getPackageIdString(pkgID));
+
```

Comment:
if `immutableTypeMap` is null. we can return here right.

Reply:
This is the better thing to do. Fixed in 48b2df122632c9fa29e166f80f23e02eedeef95c.